docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,607 @@
+	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
+}
+```
+
+### Exposing the data
+
+We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.
+
+```go
+// Hook interface used to hook into the ABCI message processing of the BaseApp
+type Hook interface {
+	ListenBeginBlock(ctx sdk.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) // update the streaming service with the latest BeginBlock messages
+	ListenEndBlock(ctx sdk.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) // update the steaming service with the latest EndBlock messages
+	ListenDeliverTx(ctx sdk.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) // update the steaming service with the latest DeliverTx messages
+}
+
+// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
+type StreamingService interface {
+	Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) // streaming service loop, awaits kv pairs and writes them to some destination stream or file
+	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register
+	Hook
+}
+```
+
+#### Writing state changes to files
+
+We will introduce an implementation of `StreamingService` which writes state changes out to files as length-prefixed protobuf encoded `StoreKVPair`s.
+This service uses the same `StoreKVPairWriteListener` for every KVStore, writing all the KV pairs from every KVStore
+out to the same files, relying on the `StoreKey` field in the `StoreKVPair` protobuf message to later distinguish the source for each pair.
+
+The file naming schema is as such:
+* After every `BeginBlock` request a new file is created with the name `block-{N}-begin`, where N is the block number. All
+subsequent state changes are written out to this file until the first `DeliverTx` request is received. At the head of these files,
+  the length-prefixed protobuf encoded `BeginBlock` request is written, and the response is written at the tail.
+* After every `DeliverTx` request a new file is created with the name `block-{N}-tx-{M}` where N is the block number and M
+is the tx number in the block (i.e. 0, 1, 2...). All subsequent state changes are written out to this file until the next
+`DeliverTx` request is received or an `EndBlock` request is received. At the head of these files, the length-prefixed protobuf
+  encoded `DeliverTx` request is written, and the response is written at the tail.
+* After every `EndBlock` request a new file is created with the name `block-{N}-end`, where N is the block number. All
+subsequent state changes are written out to this file until the next `BeginBlock` request is received. At the head of these files,
+  the length-prefixed protobuf encoded `EndBlock` request is written, and the response is written at the tail.
```

Comment:
Will this not create way too many files for the filesystem at some point. eg. 10k blocks * 100 tx = 1 million files in a dir.

I guess there can be cleanup, but why not just one file per block?

Reply:
Yeah this will make a ton of files, although perhaps one advantage of splitting a block into separate files for each ABCI request/response stage is that we can prune the files with that level of granularity. Additionally the separate files help direct our decoding of the binary data contained. E.g. for a `block-{N}-begin` file we know the first length-prefixed protobuf encoded entry is the `BeginBlock` request, that all subsequent length-prefixed entries are the resulting `StoreKVPair` state changes, until the final length-prefixed entry which is the response. If we mix, for example, `BeginBlock` and `DeliverTx` stages in the same file I think we would then need additional structure to tell us e.g. when the `BeginBlock` state changes are finished and the `BeginBlock` response begins since the `BeginBlock` response is no longer at the tail of the file. I'm nt sure that any of this is good enough justification for the additional file clutter, what do you think?

Reply:
I think one file per block is plenty for granularity. Unless you need to start processing state changes from some txs before the endblock is written, I see no downside for grouping them 1/block.

It would need a clear file-format, as you mentioned. Ideally something that could be streamed

Reply:
Increased granularity allows the sync process to work faster IMHO. The idea is really that these logs just get deleted as soon as they are consumed. It's basically a very dumb durable message queue. A single file per block could also work, we'd just need to specify the format better.